### PR TITLE
nixos/cockroachdb: simplify dataDir management, tweaks

### PR DESCRIPTION
**Pull request by thoughtpolice**
This cleans up the CockroachDB expression, with a few suggestions from
@aszlig.

However, it brought up the note of using systemd's StateDirectory=
directive, which is a nice feature for managing long-term data files,
especially for UID/GID assigned services. However, it can only manage
directories under /var/lib (for global services), so it has to introduce
a special path to make use of it at all in the case someone wants a path
at a different root.

While the dataDir directive at the NixOS level is _occasionally_ useful,
I've gone ahead and removed it for now, as this expression is so new,
and it makes the expression cleaner, while other kinks can be worked out
and people can test drive it.

CockroachDB's dataDir directive, instead, has been replaced with
systemd's StateDirectory management to place the data under
/var/lib/cockroachdb for all uses.

There's an included RequiresMountsFor= clause like usual though, so if
people want dependencies for any kind of mounted device at boot
time/before database startup, it's easy to specify using their own
mount/filesystems clause.

This can also be reverted if necessary, but, we can see if anyone ever
actually wants that later on before doing it -- it's a backwards
compatible change, anyway.

Signed-off-by: Austin Seipp <aseipp@pobox.com>

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/modules/services/databases/cockroachdb.nix [hunk 1]
```diff
@@ -193,27 +194,21 @@ in
...
 
-        unitConfig.RequiresMountsFor = "${cfg.dataDir}";
-
-        preStart = ''
-          if ! test -e ${cfg.dataDir}; then
-            mkdir -m 0700 -p ${cfg.dataDir}
-            chown -R ${cfg.user} ${cfg.dataDir}
-          fi
-        '';
+        unitConfig.RequiresMountsFor = "/var/lib/cockroachdb";
```
Comment (aszlig) [inline nixos/modules/services/databases/cockroachdb.nix]: From `systemd.exec(5)`:

> Units with WorkingDirectory=, RootDirectory=, RootImage=, RuntimeDirectory=, StateDirectory=, CacheDirectory=, LogsDirectory= or ConfigurationDirectory= set automatically gain dependencies of type Requires= and After= on all mount units required to access the specified paths. This is equivalent to having them listed explicitly in RequiresMountsFor=.

File: nixos/modules/services/databases/cockroachdb.nix [hunk 2]
```diff
@@ -140,9 +139,11 @@ in
           data for SQL queries.
 
           This can be a percentage, expressed with a fraction sign or as a
-          decimal-point number, or any bytes-based unit. For example, "25%",
-          "0.25" both represent 25% of the available system memory. The values
-          "1000000000" and "1GB" both represent 1 gigabyte of memory.
+          decimal-point number, or any bytes-based unit. For example,
+          <literal>"25%"</literal>, <literal>"0.25"</literal> both represent
```
Comment (aszlig) [inline nixos/modules/services/databases/cockroachdb.nix]: Hm, maybe remove the double quotes as well, as `<literal/>` are already formatted in a different way in the manual.

File: nixos/modules/services/databases/cockroachdb.nix [hunk 3]
```diff
@@ -70,10 +70,12 @@ in
           like datacenter.  The tiers and order must be the same on all nodes.
           Including more tiers is better than including fewer. For example:
 
+          <literal>
               country=us,region=us-west,datacenter=us-west-1b,rack=12
               country=ca,region=ca-east,datacenter=ca-east-2,rack=4
 
               planet=earth,province=manitoba,colo=secondary,power=3
+          </literal>
```
Comment (aszlig) [inline nixos/modules/services/databases/cockroachdb.nix]: Wouldn't that be one single line? Like eg. if used for the example attribute of `mkOption`:

```nix
mkOption {
  type = types.nullOr types.str;
  default = null;
  example = "country=us,region=us-west,datacenter=us-west-1b,rack=12,"
          + "country=ca,region=ca-east,datacenter=ca-east-2,rack=4,"
          + "planet=earth,province=manitoba,colo=secondary,power=3";
  description = ''
    ...
  '';
}
```

File: nixos/modules/services/databases/cockroachdb.nix (hunk 2 above)
Comment (thoughtpolice) [reply nixos/modules/services/databases/cockroachdb.nix]: Do they actually put quotes, though? I should have double checked, I just left them to make it a bit more clear they're of `string` type.

File: nixos/modules/services/databases/cockroachdb.nix (hunk 3 above)
Comment (thoughtpolice) [reply nixos/modules/services/databases/cockroachdb.nix]: Ah, maybe. I'll double check this too. `<example>` doesn't seem to work though (some kind of nested `<para>` conflict, I think?) but I'll look around for this, too.

Comment (aszlig) [reply nixos/modules/services/databases/cockroachdb.nix]: [`<example/>`](https://tdg.docbook.org/tdg/5.0/example.html) work within `<para/>`, but you'll need a `<title/>` or `<info/>` child. [`<informalexample/>`](https://tdg.docbook.org/tdg/5.0/informalexample.html) might be better suited here. Or as said: It's maybe even better to use the `example` attribute of `mkOption`.

**Comment by Infinisil** [comment]
@thoughtpolice Can you wait until people have reviewed your PRs before merging them?

**Pull request merged**